src/main/java/ch/njol/skript/util/Contract.java

```diff
@@ -0,0 +1,44 @@
+/**
+ *   This file is part of Skript.
+ *
+ *  Skript is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  Skript is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with Skript.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Copyright Peter Güttinger, SkriptLang team and contributors
+ */
+package ch.njol.skript.util;
+
+import ch.njol.skript.lang.Expression;
+import org.eclipse.jdt.annotation.Nullable;
+
+/**
+ * The 'contract' of a function or another callable.
+ * This is a non-exhaustive helper for type hints, singularity, etc. that may change based on the arguments
+ * passed to a callable, in order for it to make better judgements on correct use at parse time.
+ */
+public interface Contract {
```

Comment:
do you think this is worth generifying? mainly for return types.

Reply:
We could do, but I'm not a huge fan of type parameters where they're not providing value.

I can't think of a situation where we would actually know the generic type of this and be able to use it _without an unchecked cast_.
Even if we got the type from the function and returned it (in the function call expression) we'd still need to perform an unchecked cast for the function return because with (for example) `Function<A>` and `Contract<B>` the scope of the reified parameters probably won't be the same.